### ISSUE TITLE
Add Weights to ML model, then modify code for creating training set to include weights, -1 only used when action directly

### ISSUE THREAD

**Issue by sevin7352** [sevin7352/BlackJackTrainner#38]
-1 only used when action directly resulted in loss. (busted?) or should it be -1 when move was not optimal based on single layer analyses 

Yes! ML.NET supports instance weights, which allow you to give more importance to specific samples in your training data. In your case, you can use a weight column that assigns values based on the outcome of the action.

How to Include Weights in Your Training Pipeline
Add a new column (Weight) to your training data.
Assign weights based on the game outcome:
1.0 → Action led to a win
0.0 → Push (neutral outcome)
-1.0 → Action directly resulted in a loss
Modify your pipeline to include weightColumnName: "Weight".
Updated Pipeline with Weights
csharp
Copy
Edit
return mlContext.Transforms.Conversion.MapValueToKey("Label", "Action") // Convert Action to categorical key and rename it to "Label"
        .Append(mlContext.Transforms.Conversion.ConvertType("PlayerHandSum", "PlayerHandSum", DataKind.Single))
        .Append(mlContext.Transforms.Conversion.ConvertType("DealerUpCard", "DealerUpCard", DataKind.Single))
        .Append(mlContext.Transforms.Conversion.ConvertType("NumberOfAces", "NumberOfAces", DataKind.Single))
        .Append(mlContext.Transforms.Conversion.ConvertType("CanSplit", "CanSplit", DataKind.Single))
        .Append(mlContext.Transforms.Conversion.ConvertType("CanDouble", "CanDouble", DataKind.Single))
        .Append(mlContext.Transforms.Conversion.ConvertType("DeckCount", "DeckCount", DataKind.Single))
        .Append(mlContext.Transforms.Conversion.ConvertType("Weight", "Weight", DataKind.Single)) // Convert weight to float
        .Append(mlContext.Transforms.Concatenate("Features", "PlayerHandSum", "DealerUpCard", "NumberOfAces", "CanSplit", "CanDouble","DeckCount"))
        .Append(mlContext.MulticlassClassification.Trainers.OneVersusAll(
            mlContext.BinaryClassification.Trainers.FastTree(
                numberOfLeaves: 50, 
                numberOfTrees: 200, 
                learningRate: 0.1
            ), labelColumnName: "Label", weightColumnName: "Weight")) // Use weights in training
        .Append(mlContext.Transforms.Conversion.MapKeyToValue("PredictedAction", "PredictedLabel")); // Convert back to original label values
How to Assign Weights in Your Training Data
When preparing your dataset, make sure to include a Weight column:

csharp
Copy
Edit
var trainingData = new List<BlackJackTrainingRecord>
{
    new BlackJackTrainingRecord { Action = "Hit", PlayerHandSum = 15, DealerUpCard = 10, Weight = -1 },
    new BlackJackTrainingRecord { Action = "Stand", PlayerHandSum = 20, DealerUpCard = 10, Weight = 1 },
    new BlackJackTrainingRecord { Action = "Double", PlayerHandSum = 11, DealerUpCard = 6, Weight = 1 },
    new BlackJackTrainingRecord { Action = "Split", PlayerHandSum = 8, DealerUpCard = 5, Weight = 0 }
};
This approach ensures that the model prioritizes actions that led to wins while discouraging losing moves. Let me know if you need further refinements! 🚀

**Comment by sevin7352** [comment]
not possible with Data Classification.  only available with regression.
